src/pydata_sphinx_theme/toctree.py

```diff
@@ -36,7 +37,25 @@ def add_toctree_functions(
     """Add functions so Jinja templates can add toctree objects."""
 
     @lru_cache(maxsize=None)
-    def generate_header_nav_html(n_links_before_dropdown: int = 5) -> str:
+    def get_id_generator(base_id: str) -> Iterator[str]:
```

Comment:
slightly better name IMO would be `html_id_generator`

Reply:
So it's really more like "create_or_get"

Reply:
for Python devs, "generator" means "iterator that yields single entries on demand / doesn't precompute/store all its entries".  It doesn't imply unique values each time it's called. IDK about other python devs, but prefixing with "get_" does not unambiguously communicate "memoized" to me.

In other news, is the output type hint correct here? I don't regularly use type hints in my other projects 🙈 what is yielded is always `str` (never a list of str, tuple of str, etc)

Reply:
Totally with you on "get" not communicating "memoized." In general, I don't like trying to convey any subtlety via a function name, so I'm glad you called my questionable choice into question. Maybe the function name should be `get_or_create_html_id_generator`? A bit verbose—but clear—I think.

> doesn't imply unique values each time it's called

My question about `_generator()` function was not about the generator returning unique values but about the generator instance being unique. I think that the functions in the Python standard library follow this rule, for example:

```
itertools.count() != itertools.count()
```

Whereas my generator function:

```
html_id_generator() == html_id_generator()
```

None of this is really an issue here given that this function isn't being used in a lot of different places. Asking more from a place of wanting to understand, learn and grow.

Reply:
As for the type hint, I got that from somewhere off the Internet, don't remember where. I think I read that it's acceptable to use `Iterator` in the type hint for the return value of a generator function.

Reply:
> Maybe the function name should be get_or_create_html_id_generator? A bit verbose—but clear—I think.

To me it is misleading to call it "get_..._generator" because it implies that it returns a generator (it does not; it returns a string).

>  I think I read that it's acceptable to use Iterator in the type hint for the return value of a generator function

It's correct if the function *returns* a generator. It's incorrect (but usually not detected by type checkers! --- because strings are iterable) if the function *is* a generator and *returns* a string.

Reply:
As I was revisiting this PR today, I realized that I'm not sure we ever resolved this discussion. 

In the most up to date version of this PR, the function looks like this:

```py
@lru_cache(maxsize=None)
def get_or_create_id_generator(base_id: str) -> Iterator[str]:
    for n in count(start=1):
        if n == 1:
            yield base_id
        else:
            yield f"{base_id}-{n}"
```

I'm totally confused by your previous remarks (about this function returning a string, not a generator). The following Python REPL may help explain why:

```python-console
>>> from itertools import count
>>> def get_or_create_id_generator(base_id):
...     for n in count(start=1):
...         if n == 1:
...             yield base_id
...         else:
...             yield f"{base_id}-{n}"
... 
>>> maybe_generator = get_or_create_id_generator("foo")
>>> type(maybe_generator)
<class 'generator'>
>>> type("foo")
<class 'str'>
>>> next(maybe_generator)
'foo'
>>> next(maybe_generator)
'foo-2'
```

As the REPL log shows, when I call a function that is nearly identical to the one I wrote and check the type of the return value, I get `<class 'generator'>`, which suggests to me that the return type is a generator not a string. Furthermore, when I pass the return value to `next()` I get the next string value in a sequence, which suggests to me that the thing returned by my `get_or_create_id_generator` is something that follows the iterator protocol and therefore can be typed with `Iterator[str]` ... unless I have misunderstood something along the way here, which is totally possible ... Does the `@lru_cache` decorator change things?

Reply:
you have not misunderstood anything. I was simply wrong / looking too quickly / not thinking clearly when I said those things. Sorry to have sent you down a rabbit hole.

Reply:
Ah no worries! I didn't mind the rabbit hole :D